CRS-384 Upgrade react-virtuoso to v1

ath92 opened the pull request:
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

## Description of the pull request

Upgrades react-virtuoso to the latest version available. A refactor of `VirtualizedMessageList` was required to adapt to the breaking changes introduced with this new version. In particular, some extra work was needed to handle the new way virtuoso deals with prepending messages. 

Improvements:
- VirtualMessageList works well with `Message` and `FixedHeightMessage`
- Margins in the item content are handled
- New reactions keep the list at the bottom (if already there)

Not included:
- Images with no specified sizing displace the timeline. The same happens in non-virtualized mode.

Squash the commits before merging, a lot of back and forth.

DanC5 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -108,15 +104,46 @@ const VirtualizedMessageList = ({
     },
-    [MessageDeleted, customMessageRenderer, shouldGroupByUser],
+    [
+      MessageDeleted,
```
can probably remove MessageDeleted from dep array, it shouldn't change

ath92 replied:
Not sure if I'd remove this. Even if it shouldn't change this, it technically could, and it doesn't really hurt to have it in the dependencies array.

DanC5 replied:
I think this raises a larger question of how we handle dependency arrays. On the react native side we only include the data that should actually trigger a new render during an app session, so we don't include things like prop UI components that don't change mid-session. Obviously native is a little different, but we definitely saw performance improvements and decreased re-renders upon getting stricter with the dependency arrays. @mahboubii do you have thoughts on whether or not to keep UI components in dependency arrays?

mahboubii replied:
There could be customers who change the UI props conditionally in their app but the performance gain from removing this dependency check is worth it IMO and we should move toward that path. 
In the meantime since changing these effect dependencies are potentially a breaking change, let's leave it as is until the next major release.
we should make it clear for customers that UI prop overrides should not change after the first render of the app

ath92 replied:
Yeah, I added this to the component:
```
// TODO: split scrollSeekPlaceholder into two props (e.g. ScrollSeekPlaceholder and scrollSeekConfiguration) when making breaking changes
```

and this to the typedefs:
```
*  Note: virtuoso has broken out the placeholder value and instead includes it in its components prop.
*  TODO: break out placeholder when making other breaking changes.
```

DanC5 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -108,15 +104,46 @@ const VirtualizedMessageList = ({
+      Footer,
+    };
+  }, [
+    EmptyStateIndicator,
```
same with EmptyState and Typing

DanC5 commented:
@ath92 I'll admit I don't have much background on this feature but it looks pretty good, will wait for Amin to check it out too. Though in the meantime can you replace @ts-ignore with @ts-expect-error? This will help us when we get to converting this file and will notify you if there's not a type error to follow.

DanC5 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -78,23 +79,18 @@ const VirtualizedMessageList = ({
       // use custom renderer supplied by client if present and skip the rest
-      if (customMessageRenderer) return customMessageRenderer(messageList, i);
+      if (customMessageRenderer)
+        return customMessageRenderer(messageList, streamMessagesIndex);
```
lets wrap in brackets when an inline return gets newlined

mahboubii commented:
looks like the new `scrollTo` function also accepts `smooth` behavior

ath92 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -125,27 +152,14 @@ const VirtualizedMessageList = ({
-        }
+        itemContent={(i) => messageRenderer(messages, i)}
+        components={virtuosoComponents}
+        firstItemIndex={-numItemsPrepended}
```
This one's a bit funny, I'm not sure what's the best way to approach this. The virtuoso examples for this all have a large, mostly empty array as the initial state. As more items are backfilled (i.e. our loadMore function), the examples add these prepended items to the empty array, and the `firstItemIndex` is set to the first index in this mostly empty array that has actual messages. See https://virtuoso.dev/prepend-items/

I'd rather not keep a separate copy of our `messages` array with a bunch of empty slots at the start as state here, that feels very wasteful. I can increase the `firstItemIndex` by some arbitrary amount `x`, only to subtract `x` from the index when we use it to access the `messages` array, but I don't know why that's needed aside from the warning in the virtuoso docs.

To clarify, here's what I'm suggesting we could do:
```
const prependOffset = 10000;

// ....


  const messageRenderer = useCallback(
    (messageList, virtuosoIndex) => {
      const streamMessagesIndex = virtuosoIndex + numItemsPrepended - prependOffset;

// ....

      <Virtuoso
        itemContent={(i) => messageRenderer(messages, i)}
        components={virtuosoComponents}
        firstItemIndex={prependOffset - numItemsPrepended}

```
This would keep the index above 0 for the first 10000 messages that are preloaded. But I'm not sure what the difference would be; it seems to work exactly the same way as the current code.

mahboubii replied:
Interesting, I wonder why it specifically says to use positive values here, maybe there are some gotchas we're missing.
but having sth like `const [firstItemIndex, setFirstItemIndex] = useState(50000);` and reducing this number everytime we load more should be good for now

petyosi replied:
Sorry for the inconvenience on that. Internal boundaries are affected by a negative firstItemIndex :(. I will give this some thought, but in the meantime, using an arbitrary large number should work fine.

mahboubii replied:
btw great job on v1 @petyosi 👍

ath92 replied:
For sure! I'm not against using a positive index, it just wasn't quite clear to me from the docs why it's needed. I'd love to get a deeper understanding of Virtuoso btw, looks like there's some cool stuff going on under the hood but I find it a little difficult to dig through the source code at the moment because I've never done much with rx-like architectures.

ath92 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -125,27 +152,14 @@ const VirtualizedMessageList = ({
         overscan={overscan}
         followOutput={true}
-        maxHeightCacheSize={2000} // reset the cache once it reaches 2k
-        scrollSeek={scrollSeekPlaceHolder}
```
Hmm, so I updated the `components` prop to include the `scrollSeekPlaceholder`, but the `ScrollSeekConfiguration` is new. Should we allow SDK users to override this virtuoso prop? Should we provide a default value?

mahboubii replied:
we don't have a default one for now, I'd just check and break the user passed prop into UI and config until we go with a major release and update the API

ath92 replied:
I've updated the code now, thanks for pointing this out because I didn't realise there was so much going on in that prop. Somehow the typescript definitions also got a little messed up so tsc didn't complain that I missed a bunch of stuff there. Should be better now, but I do think the code looks a bit ugly. Definitely something we'll want to fix the next time we make breaking changes.

mahboubii replied:
awesome thanks, can you also leave a todo comment so we won't forget about it later

ath92 replied:
Yeah, I added this to the component:
```
// TODO: split scrollSeekPlaceholder into two props (e.g. ScrollSeekPlaceholder and scrollSeekConfiguration) when making breaking changes
```

and this to the typedefs:
```
*  Note: virtuoso has broken out the placeholder value and instead includes it in its components prop.
*  TODO: break out placeholder when making other breaking changes.
```

ath92 commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -125,27 +152,14 @@ const VirtualizedMessageList = ({
         totalCount={messages.length}
         overscan={overscan}
         followOutput={true}
```
`alignToBottom` seems nice, but it's not consistent with our other messagelists?

ath92 replied:
Also, `smooth` would be cool in most cases but the virtuoso docs mention that it may not be able to keep up with many messages coming in quickly. Might be risky for livestream customers?

mahboubii replied:
we can put in the default and allow customers to override them if it's not working for them

ath92 commented:
I think I've addressed the comments in this review so far, but I'm running into an issue: sometimes the `startReached` callback is called twice (one after the other, so after the first promise resolves, the function is called again to load more items). This causes the list to jump to the latest set of items that has been loaded (because `firstItemIndex` is changed when the last set of messages is loaded).

I'm not sure why this happens - as far as I can tell we're not doing anything different from the examples provided on the Virtuoso website. Perhaps @petyosi has some idea on why this could be happening? (let me know if it makes more sense to try and create a more minimal example of the problem and open an issue in the virtuoso repo instead)

petyosi commented:
> I think I've addressed the comments in this review so far, but I'm running into an issue: sometimes the `startReached` callback is called twice (one after the other, so after the first promise resolves, the function is called again to load more items). This causes the list to jump to the latest set of items that has been loaded (because `firstItemIndex` is changed when the last set of messages is loaded).
> 
> I'm not sure why this happens - as far as I can tell we're not doing anything different from the examples provided on the Virtuoso website. Perhaps @petyosi has some idea on why this could be happening? (let me know if it makes more sense to try and create a more minimal example of the problem and open an issue in the virtuoso repo instead)

I have been struggling with this problem in this issue: https://github.com/petyosi/react-virtuoso/issues/281 - check if debouncing the callback handle won't resolve it. If not, a repro will help me. Notice that the OS matters - if you reproduce it on Mac, then you are probably facing something different.

ath92 commented:
> I have been struggling with this problem in this issue: [petyosi/react-virtuoso#281](https://github.com/petyosi/react-virtuoso/issues/281) - check if debouncing the callback handle won't resolve it. If not, a repro will help me. Notice that the OS matters - if you reproduce it on Mac, then you are probably facing something different.

Got some time to look into this again: The issue I was seeing was unrelated to the issue you mentioned here (and unrelated to virtuoso), but caused by how we handle prepending items to our channel state. 

In short: the promise that is returned by `loadMore` contains the amount of new messages, but those new messages aren't added to the channel state at exactly the moment the promise resolves: the actual state update is performed by a debounced `loadMoreFinished` function in the Channel component. In some cases, this caused the list to be re-rendered with new messages, while `firstItemIndex` was not yet updated, or the other way around (`firstItemIndex` already updated, but no new messages yet). 

I pushed a workaround for this that doesn't depend on the timing of the `loadMore` promise resolving, but I'm not super happy with it because it's kinda complex. I don't immediately see an easier way to fix it though, but open to suggestions.

petyosi commented:
@mahboubii @DanC5 @virdesai please take a look

mahboubii commented on src/components/MessageList/hooks/usePrependMessagesCount.js:
```diff
@@ -0,0 +1,44 @@
+ * @param {Array<{ id: string }> | undefined} messages,
+ */
+export const usePrependedMessagesCount = (messages) => {
+  const firstMessageId = useRef(messages && messages.length && messages[0].id);
```
```suggestion
  const firstMessageId = useRef(messages?.[0]?.id);
```

mahboubii commented on src/components/MessageList/hooks/usePrependMessagesCount.js:
```diff
@@ -0,0 +1,44 @@
+ */
+export const usePrependedMessagesCount = (messages) => {
+  const firstMessageId = useRef(messages && messages.length && messages[0].id);
+  const earliestMessageId = useRef(
+    messages && messages.length && messages[0].id,
+  );
```
```suggestion
  const earliestMessageId = useRef(messages?.[0]?.id);
```

mahboubii commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -1,27 +1,24 @@
+import { useShouldForceScrollToBottom } from './hooks/useShouldForceScrollToBottom';
+import MessageNotification from './MessageNotification';
+
+const prependOffset = 10000;
```
should we increase this to 50k just to be on the safe side? what happens if we prepend more than 10k?

petyosi commented:
@DanC5 let's give this a review when possible

DanC5 commented on examples/messaging/src/App.js:
```diff
@@ -10,7 +10,7 @@ import {
   MessageInput,
   MessageInputFlat,
-  MessageList,
+  MessageList as TheMessageList,
```
do we need this name reassignment?

petyosi replied:
No, not really, it was a convenience for testing. I reverted the messaging app tweaks.

DanC5 replied:
great looks good, will merge

kainosnoema commented on src/components/MessageList/VirtualizedMessageList.js:
```diff
@@ -32,91 +29,117 @@ const VirtualizedMessageList = ({
+  // TODO: refactor to scrollSeekPlaceHolderConfiguration and components.ScrollSeekPlaceholder, like the Virtuoso Component
   scrollSeekPlaceHolder,
-  Message = FixedHeightMessage,
+  Message = DefaultMessage,
```
Seeing that a fixed height message might not be required anymore, is it possible to use the `MessageTeam` component in the virtualized list now?

Looks like this Message property requires a type with `FixedHeightMessageProps` still, so not sure how to do it...

DanC5 replied:
@petyosi can you speak to the feasibility of using our other Message UI components in the VL now?

petyosi replied:
@kainosnoema indeed, using "any" messages in the virtual list is something which we are looking into, but not yet ready to introduce as the default configuration. The part which does not work as smoothly as we wish is messages which contain `img` elements without a specified size. Such posts may cause the virtual list to "jump" when the image is loaded.

This being said, you can ignore the typing for now and try passing the plain message - we would love to get your feedback on the performance of that.

kainosnoema replied:
@petyosi thanks for the update. We're building our own custom message component and will be having photo galleries with a fixed height (when attachments are present), so don't think we'll run into the jumping issue. We'll give this a shot!
